Makefile

```diff
@@ -8,268 +28,417 @@ usage:
+	@echo "         $$(tput bold)prepush$$(tput sgr0): Fast checks to run before pushing changes upstream"
+	@echo "          ci-all: Run all continuous integration tests (possibly slow!)"
+	@echo "         ci-help: More information on Tock CI and testing"
 	@echo
 	@echo "$$(tput bold)Happy Hacking!$$(tput sgr0)"
 
+##
+## End: usage.
+##
+################################################################################
+##
+## Utility functions
+##
+
+define banner
+	@printf "\n"
+	@printf "$$(tput bold)********************************************************************************$$(tput sgr0)\n"
+	@string="$(1)" && printf "$$(tput bold)* %-$$((76))s *\n" "$$string"
+	@printf "$$(tput bold)********************************************************************************$$(tput sgr0)\n"
+	@printf "\n"
+endef
+
+##
+## End: functions.
+##
+################################################################################
+##
+## User convenience targets
+##
+
+# Aggregate targets
+.PHONY: allaudit
+audit:
+	@for f in `./tools/list_lock.sh`;\
+		do echo "$$(tput bold)Auditing $$f";\
+		(cd "$$f" && cargo audit || exit 1);\
+		done
+
 .PHONY: allboards
 allboards:
-	@for f in `./tools/list_boards.sh`; do echo "$$(tput bold)Build $$f"; $(MAKE) -C "boards/$$f" || exit 1; done
+	@for f in $(ALL_BOARDS);\
+		do echo "$$(tput bold)Build $$f";\
+		$(MAKE) -C "boards/$$f" || exit 1;\
+		done
 
 .PHONY: allcheck
 allcheck:
-	@for f in `./tools/list_boards.sh`; do echo "$$(tput bold)Check $$f"; $(MAKE) -C "boards/$$f" check || exit 1; done
+	cargo check
 
 .PHONY: alldoc
 alldoc:
-	@for f in `./tools/list_boards.sh`; do echo "$$(tput bold)Documenting $$f"; $(MAKE) -C "boards/$$f" doc || exit 1; done
+	@for f in $(ALL_BOARDS);\
+		do echo "$$(tput bold)Documenting $$f";\
+		$(MAKE) -C "boards/$$f" doc || exit 1;\
+		done
 
 
+# Commands
+.PHONY: fmt format
+fmt format: tools/.format_fresh
+	$(call banner,Formatting complete)
 
-###################################################################
+# Get a list of all rust source files (everything fmt operates on)
+$(eval RUST_FILES_IN_TREE := $(shell git ls-files | grep '\.rs$$'))
+tools/.format_fresh: $(RUST_FILES_IN_TREE)
+	@./tools/run_cargo_fmt.sh $(TOCK_FORMAT_MODE)
+	@touch tools/.format_fresh
+
+.PHONY: clean
+clean:
+	@echo "$$(tput bold)Clean top-level Cargo workspace" && cargo clean
+	@for f in `./tools/list_tools.sh`;\
+		do echo "$$(tput bold)Clean tools/$$f";\
+		cargo clean --manifest-path "tools/$$f/Cargo.toml" || exit 1;\
+		done
+	@echo "$$(tput bold)Clean rustdoc" && rm -rf doc/rustdoc
+	@echo "$$(tput bold)Clean ci-artifacts" && rm -rf ./ci-artifacts
+
+
+## Meta-Targets
+
+# Run all possible CI. If this passses locally, all cloud CI *must* pass as well.
+.PHONY: ci-all
+ci-all:\
+	ci-runner-github\
+	ci-runner-netlify
+
+# Run the fast jobs.
+# This is designed for developers, to be run often and before submitting code upstream.
+.PHONY: prepush
+prepush:\
```

Comment:
As a developer, I'd still like to have a rule that builds boards and runs checks/tests locally, without having to upload my work-in-progress code to a pull request to get feedback from the GitHub actions. Maybe in the "hidden convenience targets" section?

Reply:
Does `make ci-all` cover that need for you? Or is there something else you have in mind?

Reply:
I've already mentioned it in https://github.com/tock/tock/pull/1861 (in particular https://github.com/tock/tock/pull/1861#issuecomment-630982811), which started all this discussion.

Specifically, I've already mentioned (https://github.com/tock/tock/issues/1865#issuecomment-630976331) that the following chain is highly problematic on a development machine.

```
ci-all
    -> ci-runner-github
    -> ci-runner-github-qemu
    -> ci-job-qemu
    -> ci-setup-qemu
    -> git clone https://github.com/alistair23/qemu.git && ./configure
    -> execution of arbitrary code without any integrity check from Tock
```

The currently proposed `make prepush` is fine and fast, but not that useful as it only checks code formatting and simple lints. I'd expect at the very least a rule that also includes `make allboards` (so that boards indeed compile). And more generally all the rules that only depend on Cargo.

One could do `make prepush && make allboards && make foobar`, but I'd be nice to simply have a rule for that (like there is the "hidden" `make clippy` rule).

Reply:
OH! I messed that up! That setup rule should prompt in non-CI context. I'll update that, sorry!

Reply:
Okay, this is now fixed. This `ci-all` rule will now prompt and offer to do the install for you. There is also a new `ci-nosetup` rule that runs all the CI that requires no new installation on the machine it's being run on.